### PR TITLE
New `Mod` class for SymPy expressions.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104968

This PR introduces a new `Mod` class to be used with SymPy expressions. The main reason
being due to SymPy simplification errors (#97792).

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: test/test_proxy_tensor.py [hunk 1]
```diff
@@ -958,6 +958,8 @@ def f(x):
         self.assertFalse(eval_guards(gm, torch.randn(25, 5)))
         self.assertExpectedInline(show_guards(gm), """L['x'].size()[0] < 20""")
 
+    # Relies on simplification of: Ne(Mod(s0, s0*s2), 0)
```
Comment (ezyang) [inline test/test_proxy_tensor.py:961]: These should be relatively simple to do (or is this simplification unsound, if s2 is 0?)

**Review by ezyang — APPROVED** [review]
Thanks!

File: test/dynamo/test_export.py [hunk 2]
```diff
@@ -3000,6 +3000,7 @@ def forward(self, x):
     return pytree.tree_unflatten([slice_tensor, slice_tensor_3, slice_tensor_6, slice_tensor_9], self._out_spec)""",
         )
 
+    @unittest.expectedFailure
```
[the remark this answers is not in the document]
Comment (ysiraichi) [reply test/dynamo/test_export.py:3003]: We were discussing about it here: #104826

File: test/test_proxy_tensor.py (hunk 1 above)
Comment (ysiraichi) [reply test/test_proxy_tensor.py:961]: I think this is sound. But, then, we would be relying on SymPy simplification rules.
For example, I could do something like:

```python
class Mod(...):
    @classmethod
    def eval(cls, p, q):
        ...
        r = p / q
        if r.is_integer:
            return S.Zero
```

Question: is this ok to rely on?

**Comment by ysiraichi** [comment]
I have added 2 simplifications: given `Mod(p, q)`:

- Return 0 if `p / q` is an integer
- Return `p` if `q / p` is an integer and they both have the same sign

This fixes all the XFAIL I had added. However, it relies on SymPy div simplification.

File: test/dynamo/test_export.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ysiraichi) [reply test/dynamo/test_export.py]: See [this comment](https://github.com/pytorch/pytorch/pull/104968#issuecomment-1634877881).

File: test/test_proxy_tensor.py (hunk 1 above)
Comment (ysiraichi) [reply test/test_proxy_tensor.py]: See [this comment](https://github.com/pytorch/pytorch/pull/104968#issuecomment-1634877881).

**Pull request closed without merging**